### PR TITLE
Use hash of enclosingType if it is a ParameterizedTypeBinding

### PR DESCRIPTION
**Pull request by brychcy**
Fixes "Increased compilation time with Eclipse 4.12+" #549

### CODE REVIEW

File: org.eclipse.jdt.core/compiler/org/eclipse/jdt/internal/compiler/lookup/TypeSystem.java [hunk 1]
```diff
@@ -98,6 +98,9 @@ public void swapUnresolved(UnresolvedReferenceBinding unresolvedType, ReferenceB
 					ReferenceBinding enclosing = resolvedType.enclosingType();
 					if (enclosing != null) {
 						this.enclosingType = resolvedType.isStatic() ? enclosing : (ReferenceBinding) env.convertUnresolvedBinaryToRawType(enclosing); // needed when binding unresolved member type
+						if(this.enclosingType.getClass() == ParameterizedTypeBinding.class) {
```
Comment (iloveeclipse) [inline org.eclipse.jdt.core/compiler/org/eclipse/jdt/internal/compiler/lookup/TypeSystem.java:101]: May I ask for a "Ctrl + F" on selection? The code change doesn't look aligned with other formatting.

**Review by iloveeclipse — COMMENTED** [review]
Till, many thanks for the patch. Do you think a regression test is possible?

File: org.eclipse.jdt.core/compiler/org/eclipse/jdt/internal/compiler/lookup/TypeSystem.java [hunk 2]
```diff
@@ -123,6 +126,11 @@ final int hash(TypeBinding b) {
 			public int hashCode() {
 				final int prime=31;
 				int hashCode = 1 + hash(this.type);
+				if (this.enclosingType != null && this.enclosingType.getClass() == ParameterizedTypeBinding.class) {
```
Comment (eric-milles) [inline org.eclipse.jdt.core/compiler/org/eclipse/jdt/internal/compiler/lookup/TypeSystem.java:129]: Would "this.enclosingType != null && this.enclosingType.isParameterizedType()" cover it?  Or do you also need this for `RawTypeBinding`?

File: org.eclipse.jdt.core/compiler/org/eclipse/jdt/internal/compiler/lookup/TypeSystem.java (hunk 1 above)
Comment (brychcy) [reply org.eclipse.jdt.core/compiler/org/eclipse/jdt/internal/compiler/lookup/TypeSystem.java:101]: done.

File: org.eclipse.jdt.core/compiler/org/eclipse/jdt/internal/compiler/lookup/TypeSystem.java (hunk 2 above)
Comment (brychcy) [reply org.eclipse.jdt.core/compiler/org/eclipse/jdt/internal/compiler/lookup/TypeSystem.java:129]: So the base issue is that RawTypeBinding extends PTB and  raw instances can appear during URB resolution, so we cannot use "instanceof PTB".

And you're not wrong: isParameterizedType() would actually work here, but you have to hunt down all overloads to see this: TypeBinding defines it to returns false; it is overloaded to return true in PTB and then to return false again in RTB - and there are no other subclasses of PTB. 

Now here "exactly for PTB instances" is really meant so I think it is better to express this as direct as possible.

Comment (eric-milles) [reply org.eclipse.jdt.core/compiler/org/eclipse/jdt/internal/compiler/lookup/TypeSystem.java:129]: Right, but isn't ".getClass() == ParameterizedTypeBinding.class" much more expensive than ".isParameterizedType()"?

Comment (brychcy) [reply org.eclipse.jdt.core/compiler/org/eclipse/jdt/internal/compiler/lookup/TypeSystem.java:129]: No, at least hotspot optimizes this, see https://github.com/openjdk/jdk/blob/ccc6e169188019dab008df5ffb11b4279e14f8fc/src/hotspot/share/opto/subnode.cpp#L1068

**Pull request merged**